rs/registry/canister-client/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+//! An implementation of RegistryClient intended to be used in canister
+//! where polling in the background is not required because handed over to a timer.
+
+use ic_interfaces_registry::{
+    empty_zero_registry_record, RegistryClient, RegistryClientVersionedResult,
+    RegistryDataProvider, RegistryTransportRecord, ZERO_REGISTRY_VERSION,
+};
+use ic_types::{registry::RegistryClientError, RegistryVersion, Time};
+use std::collections::BTreeMap;
+use std::sync::{Arc, RwLock, RwLockReadGuard};
+
+type CacheState = (
+    RegistryVersion,
+    BTreeMap<RegistryVersion, Time>,
+    Vec<RegistryTransportRecord>,
+);
+
+pub struct CanisterRegistryClient {
+    data_provider: Arc<dyn RegistryDataProvider>,
+    cache: Arc<RwLock<CacheState>>,
+}
+
+impl CanisterRegistryClient {
+    pub fn new(data_provider: Arc<dyn RegistryDataProvider>) -> Self {
```

Comment:
Is there already a RegistryDataProvider that will work in canister context?

Reply:
Not sure in my canister I implemented one which uses `stable_structures` and it seams to be working. Once I am done with it we could move it here. wdyt?

Reply:
It's probably fine to leave it where it is, but I wouldn't object to having an implementation here for stable structures.